server/sonar-process-monitor/src/main/java/org/sonar/application/process/ProcessLauncherImpl.java

```diff
@@ -0,0 +1,290 @@
+  private static final Logger LOG = LoggerFactory.getLogger(ProcessLauncherImpl.class);
+  public static final String ELASTICSEARCH_JVM_OPTIONS_HEADER = "# This file has been automatically generated by SonarQube during startup.\n" +
+    "# Please use the sonar.search.javaOpts in sonar.properties to specify jvm options for Elasticsearch\n" +
+    "\n" +
+    "# DO NOT EDIT THIS FILE\n" +
+    "\n";
+
+  private final File tempDir;
+  private final AllProcessesCommands allProcessesCommands;
+  private final Supplier<ProcessBuilder> processBuilderSupplier;
+
+  public ProcessLauncherImpl(File tempDir) {
+    this(tempDir, new AllProcessesCommands(tempDir), JavaLangProcessBuilder::new);
+  }
+
+  ProcessLauncherImpl(File tempDir, AllProcessesCommands allProcessesCommands, Supplier<ProcessBuilder> processBuilderSupplier) {
+    this.tempDir = tempDir;
+    this.allProcessesCommands = allProcessesCommands;
+    this.processBuilderSupplier = processBuilderSupplier;
+  }
+
+  @Override
+  public void close() {
+    allProcessesCommands.close();
+  }
+
+  @Override
+  public ProcessMonitor launch(EsCommand esCommand) {
+    Process process = null;
+    ProcessId processId = esCommand.getProcessId();
+    try {
+      writeConfFiles(esCommand);
+      ProcessBuilder processBuilder = create(esCommand);
+      if (LOG.isInfoEnabled()) {
+        LOG.info("Launch process[{}]: {}", processId.getKey(), String.join(" ", processBuilder.command()));
+      }
+
+      process = processBuilder.start();
+
+      return new EsProcessMonitor(process, processId, esCommand);
+    } catch (Exception e) {
+      // just in case
+      if (process != null) {
+        process.destroyForcibly();
+      }
+      throw new IllegalStateException(format("Fail to launch process [%s]", processId.getKey()), e);
+    }
+  }
+
+  private void writeConfFiles(EsCommand esCommand) {
+    File confDir = esCommand.getConfDir();
+    if (!confDir.exists() && !confDir.mkdirs()) {
+      String error = format("Failed to create temporary configuration directory [%s]", confDir.getAbsolutePath());
+      LOG.error(error);
+      throw new IllegalStateException(error);
+    }
+
+    try {
+      IOUtils.copy(getClass().getResourceAsStream("elasticsearch.yml"), new FileOutputStream(new File(confDir, "elasticsearch.yml")));
+      IOUtils.copy(getClass().getResourceAsStream("jvm.options"), new FileOutputStream(new File(confDir, "jvm.options")));
+      esCommand.getLog4j2Properties().store(new FileOutputStream(new File(confDir, "log4j2.properties")), "log42 properties file for ES bundled in SonarQube");
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to write ES configuration files", e);
+    }
+  }
+
+  @Override
+  public ProcessMonitor launch(JavaCommand javaCommand) {
+    Process process = null;
+    ProcessId processId = javaCommand.getProcessId();
+    try {
+      ProcessCommands commands = allProcessesCommands.createAfterClean(processId.getIpcIndex());
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Use try-with-resources or close this "ProcessCommands" in a "finally" clause. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2095)